### PR TITLE
Autocomplete + Join and Leave Rooms + More Options

### PR DESCRIPTION
**Pull request by JackyChiu**
## Autocomplete
That's right boys. We got autocomplete. When tabbing you get a menu of options 😎

## Join and Leave Rooms
You can now join and leave rooms without restarting!

## More 
- Code cleanup
- Helper functions
- package and readme update

## print()
<s>Use `print()` when there's a possibility of incoming messages right after</s>

## Video
I did a quick summary of everything I did, either watch the video (starting at 1:10) or try out my branch to really experience the difference

[asciicast](https://asciinema.org/a/cnybydu5l8c3gegj77qyffx5a)

### CODE REVIEW

**Comment by JackyChiu** [comment]
## Edit
Use `print()` when you're printing something that could interrupt the user typing
ex. A incoming message

File: input.js [hunk 1]
```diff
@@ -5,13 +5,16 @@ const readline = require('readline')
...
+        output: process.stdout,
     });
 
-    rl.on('SIGINT', () => {
-        process.exit()
-    })
+    function completer(line) {
+        const completions = ':help :q :cr :gr'.split(' ')
+        const hits = completions.filter(c => c.indexOf(line) === 0)
+        return [hits.length ? hits : completions, line]
```
Comment (kirinrastogi) [inline input.js]: 💯

File: input.js [hunk 2]
```diff
@@ -5,10 +5,17 @@ const readline = require('readline')
...
+        completer,
         input: process.stdin,
-        output: process.stdout
+        output: process.stdout,
     });
 
+    function completer(line) {
+        const completions = ':help :q :cr :gr'.split(' ')
+        const hits = completions.filter(c => c.indexOf(line) === 0)
+        return [hits.length ? hits : completions, line]
```
Comment (kirinrastogi) [inline input.js:16]: 💯

**Comment by kirinrastogi** [comment]
Use of handlers is 🍆 🤓

File: input.js [hunk 3]
```diff
@@ -5,13 +5,16 @@ const readline = require('readline')
 const Input = (() => {
 
     const rl = readline.createInterface({
+        completer,
         input: process.stdin,
-        output: process.stdout
+        output: process.stdout,
```
Comment (kirinrastogi) [inline input.js]: What is the benefit of having a trailing comma here? Is this part of our style guide now?

File: input.js [hunk 4]
```diff
@@ -40,20 +43,27 @@ const Input = (() => {
...
+        readline.cursorTo(process.stdout, 0)
     }
 
     return {
         setName,
         setRoom,
         setPrompt,
         setMessage,
+        setLeave,
         clearLine,
```
Comment (kirinrastogi) [inline input.js]: Same here

**Review by kirinrastogi — CHANGES_REQUESTED** [review]
Trailing commas that are inconsistent with JSON formatting

File: README.md [hunk 5]
```diff
@@ -17,11 +17,11 @@ Everyone likes to feel `Hack3r Statu$` sometimes 😎.
...
-    node app.js
+    npm start
 
 ## Running Locally 
 
     # Head over to our server repo and clone and run it
     # Make sure in config.js, you're connecting to the local server
     # Then run
-    node app.js 
+    npm start
```
Comment (BenEmdon) [inline README.md:27]: 👉👉

File: app.js [hunk 6]
```diff
@@ -7,79 +7,134 @@ const Input = require('./input.js')
 const color = require('ansi-color').set
 
 const socket = io.connect(Config.serverURL)
-let username = 'anonymous'
 
+// User config on startup
+let username = 'anonymous'
```
Comment (BenEmdon) [inline app.js:12]: img_1844 jpg

File: app.js [hunk 7]
```diff
@@ -7,79 +7,134 @@ const Input = require('./input.js')
...
 
 socket.on('join-room', (data) => {
-    console.log(data)
+    print(`${data}\n`)
 })
+
+// Event handlers
+function messageHandler(message) {
+    switch (message) {
+        case ':q':
```
Comment (BenEmdon) [inline app.js:66]: Maybe sanitize strings before `switching` so that whitespace doesn't affect

File: input.js [hunk 8]
```diff
@@ -40,20 +43,27 @@ const Input = (() => {
...
+        rl.on('SIGINT', () => {
+            leaveHandler()
         })
     }
 
     function clearLine() {
-	readline.clearLine(process.stdout, 0)
-	readline.cursorTo(process.stdout, 0)
+        readline.clearLine(process.stdout, 0)
+        readline.cursorTo(process.stdout, 0)
```
Comment (BenEmdon) [inline input.js:58]: What are thoseee??

File: input.js (hunk 4 above)
Comment (BenEmdon) [reply input.js]: In most modern languages they're allowed but serve no purpose

File: package.json [hunk 9]
```diff
@@ -4,22 +4,22 @@
...
   "license": "MIT",
   "bugs": {
-    "url": "https://github.com/Hacker-Chatroom/Hacker-Chatroom-Client/issues"
+    "url": "https://github.com/Branch-Bunch/Hacker-Chatroom-Client/issues"
   },
-  "homepage": "https://github.com/Hacker-Chatroom/Hacker-Chatroom-Client#readme",
+  "homepage": "https://github.com/Branch-Bunch/Hacker-Chatroom-Client#readme",
   "dependencies": {
     "ansi-color": "^0.2.1",
-    "request": "^2.76.0",
```
Comment (BenEmdon) [inline package.json:22]: bye bye 👋 😢

File: input.js [hunk 10]
```diff
@@ -5,13 +5,16 @@ const readline = require('readline')
...
-        output: process.stdout
+        output: process.stdout,
     });
 
-    rl.on('SIGINT', () => {
-        process.exit()
-    })
+    function completer(line) {
+        const completions = ':help :q :cr :gr'.split(' ')
+        const hits = completions.filter(c => c.indexOf(line) === 0)
```
Comment (BenEmdon) [inline input.js:15]: so no leading whitespace?  Not familiar with vim but I suppose it also doesn't allow for leading whitespace?

File: app.js [hunk 11]
```diff
@@ -7,79 +7,134 @@ const Input = require('./input.js')
...
+
+function leaveHandler() {
+    socket.emit('leave-room', username)
+    process.exit()
+}
+
+function joinHandler(room) {
+    socket.emit('join-room', room, username)
+    console.log(`Joined room: ${room}`)
+    console.log('Type ":" and tab for help and options')
```
Comment (BenEmdon) [inline app.js:115]: Is `:` a special character?

**Review by BenEmdon — APPROVED** [review]
Changes look good to me 👍 nice cleanup very elegant ✨

File: app.js (hunk 11 above)
Comment (kirinrastogi) [reply app.js:115]: It's how the client tells the difference between a command and a message.

File: input.js (hunk 8 above)
Comment (vikrambombhi) [reply input.js:58]: so the clearLine(stream, direction) clears the stream(stdout in out case) in the direction(-1 for left of cursor, 1 for right of cursor, 0 for entire line)

Comment (vikrambombhi) [reply input.js:58]: cursorTo is just moving the cursor to the start of the line

File: input.js (hunk 3 above)
Comment (JackyChiu) [reply input.js]: Good catch 👍

File: README.md (hunk 5 above)
Comment (JackyChiu) [reply README.md:27]: @BenEmdon 👈👈

File: app.js (hunk 7 above)
Comment (JackyChiu) [reply app.js:66]: Good idea 👍

File: input.js (hunk 10 above)
Comment (JackyChiu) [reply input.js:15]: Vim allows for leading and trailing white space but I think we should limit it so easier checking

File: app.js (hunk 11 above)
Comment (JackyChiu) [reply app.js:115]: Yeah and all our options start with the `:` character

File: package.json (hunk 9 above)
Comment (JackyChiu) [reply package.json:22]: dw request promise is still there!

File: app.js (hunk 7 above)
Comment (JackyChiu) [reply app.js:66]: I'm actually going to leave this to another PR to deal with all the sanitization and validation of messages before sending

**Review by kirinrastogi — APPROVED** [review]

**Pull request merged**